### ISSUE TITLE
Decision Proposal 014 - Handling Of Union Types

### ISSUE THREAD

**Issue by JamesMBligh** [ConsumerDataStandardsAustralia/standards#14]
This decision proposal outlines a recommendation for the handling of heterogeneous objects in the JSON payloads.

Feedback is now open for this proposal. Feedback is planned to be closed on the 21st September.
[Decision Proposal 014 - Handling Of Union Types.pdf](https://github.com/ConsumerDataStandardsAustralia/open-banking/files/2386471/Decision.Proposal.014.-.Handling.Of.Union.Types.pdf)

**Comment by assem-ali** [comment]
I suggest to use a standard field to represent the sub class type (ex. $type) rather than having a convention (<parentClass>Union) that's hard to code against. 

"shapes": {
    [
        {
            "x": 10,
            "y": 10,
            "$type": "circle",
            "circle": {
                "radius": 5
            }
        },
        {
            "x": 100,
            "y": 10,
            "$type": "square",
            "square": {
                "edgeLength": 10
            }
        },
        {
            "x": 100,
            "y": 10,
            "$type": "rectangle",
            "rectangle": {
                "width": 20,
                "height": 10
            }
        }
    ]
}

**Comment by JamesMBligh** [comment]
@assem-ali, I thought about that but it precludes two unions in the same object as the $type fields would clash.

What about <name>$type?  In the example the field would then be shape$type.

-JB-

**Comment by da-banking** [comment]
The Standards Body will need to produce a standard that describes which fields are mandatory, which fields are conditional (for example a loan account may have a redraw amount), and which fields are optional for data providers to implement.

The standards will need to describe what the intent and meaning of each field is so that data providers can correctly determine what data, if any, to place in each field.

The introduction of Union Types places the onus on the Standards Body to correctly define the archetypes available, and precludes data providers introducing new ones later.

For example if a data provider wanted to bring out a new type of shape - say a rhombus - with either Option 1 or Option 2, the data provider couldn't expose it without the Standards Body doing work to permit it, and data consumers making corresponding changes. A rhombus is neither a "circle", nor a "square", nor a "rectangle".

> In this scenario there are likely to be some common fields but there will also be account type specific fields.

We feel that focusing on account types, is the wrong perspective for the Union Types. 

If the Standards Body had declared a set of "conditional features" that shapes could have - being radius, edgeLength, width, and height - then the data provider could expose the rhombus as it has an edgeLength.

In our experience, it is better to design APIs to allow for some flexibility to deal with the unanticipated, than to attempt to anticipate everything up front.

We therefore recommend that Union Types be not focused on accounts per se as is stated in the examples, but on the _features_ of an account that are of interest to data consumers.

This is a subtle distinction, but worth calling out.

For example, a term deposit account may have properties that describe the account that interest payments should be paid to. This feature could be applicable to any interest bearing account, and the population of this should therefore be left to the data provider to determine. It might be a mistake for the Standards Body to define credit interest as being only applicable to a term deposit account type.

Separately, we would draw attention to the additional complexity that Union Types bring to the payload and the data consumer in general, and would question whether that complexity is justified.

The JSON example for Option 2:

```
"data": {
  [
    {
      "x": 10,
      "y": 10,
      "shapeUnion": "circle",
      "circle": {
        "radius": 5
      }
    },
    {
      "x": 100,
      "y": 10,
      "shapeUnion": "square",
      "square": {
        "edgeLength": 10
      }
    },
    {
      "x": 100,
      "y": 10,
      "shapeUnion": "rectangle",
      "rectangle": {
        "width": 20,
        "height": 10
      }
    }
  ]
}
```

Consuming this data structure would require code like:

```
foreach(var shape in shapes) {
  var x = shape.x;
  var y = shape.y;
  var shapeUnion = shape.shapeUnion;
  var radius = null;
  var edgeLength = null;
  var width = null;
  var height = null;
  
  if(shapeUnion === "circle") {
    radius = shape.circle.radius;
  } else if(shapeUnion === "square") {
    edgeLength = shape.square.edgeLength;
  } else if(shapeUnion === "rectangle") {
    width = shape.rectangle.width;
    height = shape.rectangle.height;
  }
  
  // Do something with the shape.
}
```

However the same example without Union Types would be:

```
"data": {
  [
    {
      "x": 10,
      "y": 10,
      "radius": 5
    },
    {
      "x": 100,
      "y": 10,
      "edgeLength": 10
    },
    {
      "x": 100,
      "y": 10,
      "width": 20,
      "height": 10
    }
  ]
}
```

And the code to consume it would be:

```
foreach(var shape in shapes) {
  var x = shape.x;
  var y = shape.y;
  var radius = shape.radius;
  var edgeLength = shape.edgeLength;
  var width = shape.width;
  var height = shape.height;
  
  // Do something with the shape.
}
```

It seems that not using Union Types makes the data simpler to consume. It also makes the client code more robust in that no changes are required for the data consumer, in the scenario for the new rhombus shape (above). 

We can see how the use of Union Types sidesteps possible naming conflicts, however the Standards Body is in a position to name all properties unambiguously.

We are not discouraging the use of child objects to group closely related properties together, but we see little value in introducing child objects arbitrarily.

**Comment by tonymorris** [comment]
FWIW, you can do algebraic data types (union types) in Java and achieved the *closedness* of the ADT using a language trick. I blogged about it in 2006. http://blog.tmorris.net/posts/maybe-in-java/index.html

Here is the implementation in Java that has been used in various production systems, since the early 2000s:
http://www.functionaljava.org/javadoc/4.8/functionaljava/fj/data/Either.html

It can be done in most programming languages, but using different tricks (e.g. Scott Encoding). You don't need to use `null`.

**Comment by JamesMBligh** [comment]
I did not propose either of these variants as they both result in many conditionally mandatory fields which is then hard to enforce compliance on.  The options are trade offs between strong object typing vs the simplicity of direct access.

As most of the client languages that consume JSON in the wild have the concept of objects and polymorphism I believe that, for a regime like we are building, the advantage of strong data integrity that can be tested by simple schema is the stronger driver.

Experience in other jurisdictions and standards indicate that an abundance of optional or conditionally mandatory fields results in variations between providers that impact the data consumers adversely.

-JB-

**Comment by bazzat** [comment]
The ABA Open Banking Working Group was unable to support the recommendation in this case. A WG member commented: 

"The proposal describes a generic pattern, but it was not clear in exactly which scenario this pattern would be used (we acknowledge it provided account details as an example). We would prefer to have the conversation about how it is used in a specific scenario (e.g. let's discuss account details and how we intend to diverge from the UK spec), rather than endorse a generic approach, without a clear understanding of when it is being applied."

**Comment by TKCOBA** [comment]
COBA’s view is that the option implemented for the API standards should adhere to the API Principle 7: APIs are simple.

**Comment by Pelsurry** [comment]
We agree with the concerns raised in various comments - the proposed approach appears to require the standards to define product archetypes. Theoretically, a bank account is generic but, in practice, even the most simple of accounts can have highly variable features (even within the same bank) and can appear as hybrids (eg is a mortgage offset account a deposit account, or a feature of the loan account or something inbetween).  This becomes even more difficult when considering accounts held by businesses and seems to restrict future innovation in accounts/financial products.

We would ask that this proposal be left open for a few more days for further review of the feedback provided.

**Comment by NationalAustraliaBank** [comment]
There are good arguments throughout this thread from many different perspectives. 

Perhaps this proposal and its arguments should help us form our position on the archetypes that may be in scope. A more concrete example might draw out the subtleties that would make the choice more apparent.

The outcome of this proposal would likely be needed to define the data specifications required for the different types of **Accounts** and the different types of **Products** returned by their respective end points. 

That said, NAB also recognises the time constraints that the scheme is under, and will support Data 61's guidance on this matter. If we run into roadblocks we have an agreed versioning strategy in place.

**Comment by JamesMBligh** [comment]
I am happy to leave this proposal open for a few more days as requested.  I would also note that, if the proposal is agreed, it doesn't mean it will be used.  If no use case warrants it then it will be moot so some of the concerns could be delayed until specific payloads are under discussion.

-JB-

**Comment by DeloittePE** [comment]
The main discussion here seems to be a trade-off between a strongly typed approach using a "type" keyword versus a mixin approach. It is unlikely we would ever come to a definitive choice in this forum or the timeframe available because in the abstract, there is no inherently right or wrong answer.

We note that a number of existing API standards allow for a "type" property either as an optional or mandatory field.

- [JSON-API](http://jsonapi.org/format/#document-resource-objects) has "type" as a mandatory property for all resources (not objects)
- [JSON-LD](https://w3c.github.io/json-ld-syntax/#specifying-the-type) has an optional "@type" property.
- [HAL](https://tools.ietf.org/html/draft-kelly-json-hal-08#page-5) has "type" as an optional property.

The [Jackson](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization) serialization framework commonly used by Java programmers also relies on a type identifier for handling polymorphic classes and while the specific format is "configurable" our experience is that a "type" property in the JSON serialization works well in these cases.

We have used both these approaches in real-world APIs and have found that allowing for a combination of the two approaches is useful. We note that most programmers are more familiar with strong typing and many of them struggle with mixins (probably because they come from strongly-typed languages like Java and C#).

The examples discussed so-far in this discussion are object examples (shapes) which don't fully match the API world which operates at the more coarse-grained level of resources. 

An API payload represents one or more resources (e.g. Person) each of which is composed of one or more objects (e.g. Name, Address, Contact Details). The "type" property is useful at the resource level but is seldom useful at the object level. 

In real-world APIs we have found benefit in the JSON-API approach of having a "type" property at the resource level. 

E.g. the following representation

```
{
   "id": "5f59e20061db",
   "type": "Person",
   "name": { "first": "Ada", "family": "Lovelace"}
   "address": { "streetNo": "15", "streetName": "Elm", "streetType": "St", "locality": "Melbourne", "state": "VIC", "postcode": "3000"}
}
```
identifies the main resource as being of type "Person" but doesn't really benefit from "type" in the name or address objects, since these are identified by the keywords in the property map.

Most API resources involve a projection of an information graph into a JSON tree structure, so we often see related resources brought in as branches on the tree. In these cases, the "type" keyword identifies related resources or sub-resources.

For example Person is a resource (is accessible via a URL) and Account is a resource. In a request to get an Account resource, we may find that a Person is embedded in the response, representing the holder of the account.

```
{
   "id": "e763589e9feb",
   "type": "Account",
   "bsb": "123-456",
   "accountNumber": "123456",
   "accountOwners": [
      { "type": "Person", ... }
      { "type": "Person", ... }
   ],
   "branch": { ... }
}
```

The above structure is a projection of two resources, each accessible via calls to http://api.example.com/v1/people/5f59e20061db or http://api.example.com/v1/accounts/e763589e9feb. We won't get into the nuances of inclusions and projections here, but suffice to use this example as to where the granularity of "type" is most useful.

If "type" maps to the level of resource then what happens if we have a polymorphic resource? This often happens with collection resources where we might retrieve a list of Accounts with each member of the collection having a different object representation. That is where a subtype is useful. E.g.

``GET http://api.example.com/v1/accounts`` may return a collection:

```
"data": {
     "accounts": [
         { 
           "id": "e763589e9feb",
           "type": "account",
           "accountType": "cash",
           ... cash account specific properties ...
         },
         {
           "id": "e763589e9feb",
           "type": "account",
           "accountType": "mortgage",
           ... mortgage account specific properties ...
         }
     ]
}
```

Here, we are using a combination of strong-typing (the "type" keyword) and mixins (embedded properties). The "type" operates as an instruction to the consumer as to how to interpret the resource.

Another use of mixins is in extending a genric type (such as Person) into a more specialised type involving extended properties. A real-world example in the education sector is where a Person can operate in the roles of Student, Staff and Alumnus. Very often, the same person fills two or more of these roles simultaneously. Mixins are useful in extending the base resource to fulfil the specialised roles. E.g.:

```
{
   "id": "5f59e20061db",
   "type": "Person",
   "name": { "first": "Ada", "last": "Lovelace"}
   "address": { "streetNo": "15", "streetName": "Elm", "streetType": "St", "locality": "Melbourne", "state": "VIC", "postcode": "3000"}
   "student": {  ...  student specific properties ... },
   "staff": { ... staff specific properties ... }
}
```

In summary, a flexible API allows for resources to be composed from a common vocabulary of objects that may involve inheritence as well as composition. 

Composition is easily achieved via mixins without any requirements for specific keywords or language features. But an inheritence or specialisation hierarchy is commonly facilitated by a "type" property as found in common standards such as JSON-API, JSON-LD or HAL. 

The "type" keyword is most useful at the resource level but just adds noise and complexity at the individual object level.

**Comment by JamesMBligh** [comment]
After an extension for a few daysI have now closed consultation on this decision. A recommendation incorporating feedback will be made to the Chair in due course.

-JB-

**Comment by JamesMBligh** [comment]
The finalised decision for this topic has been endorsed. Please refer to the attached document.
[Decision 014 - Handling Of Union Types.pdf](https://github.com/ConsumerDataStandardsAustralia/standards/files/2444682/Decision.014.-.Handling.Of.Union.Types.pdf)

-JB-
